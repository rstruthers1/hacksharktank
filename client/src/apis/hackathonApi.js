// hackathonApi.js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonApi = createApi({
    reducerPath: 'hackathonApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        createHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: 'hackathons', // Your endpoint path
                method: 'POST',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
        }),
        updateHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: `hackathons/${hackathonData.id}`, // Your endpoint path
                method: 'PUT',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'Hackathons'},
                { type: 'Hackathon', id: arg.id }],
        }),
        getHackathons: builder.query({
            query: () => ({
                url: 'hackathons', // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathons'}],
        }),
        getHackathon: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathon', id: arg }],
        }),
        createHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: 'hackathons/users/roles', // Your endpoint path
                method: 'POST',
                body: hackathonUserRoleData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        deleteHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: `hackathons/${hackathonUserRoleData.hackathonId}/users/${hackathonUserRoleData.userId}/roles/${hackathonUserRoleData.roleName}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        getHackathonUsers: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/users`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg }],
        }),
        deleteHackathonUser: builder.mutation({
            query: (hackathonUserData) => ({
                url: `hackathons/${hackathonUserData.hackathonId}/users/${hackathonUserData.userId}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        // You can add more endpoints here
    }),
});

export const {
    useCreateHackathonMutation,
    useUpdateHackathonMutation,
    useGetHackathonsQuery,
    useCreateHackathonUserRoleMutation,
    useGetHackathonQuery,
    useGetHackathonUsersQuery,
    useDeleteHackathonUserRoleMutation,
    useDeleteHackathonUserMutation,
} = hackathonApi;
