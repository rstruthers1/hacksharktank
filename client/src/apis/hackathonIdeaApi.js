import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonIdeaApi = createApi({
    reducerPath: 'hackathonIdeaApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        getHackathonIdeas: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/ideas`,
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
       createHackathonIdea: builder.mutation({
                query: (hackathonIdeaData) => ({
                    url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                    method: 'POST',
                    body: hackathonIdeaData.data,
                    // Add a JWT token to the request headers if the user is logged in
                    headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
                }),
                invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
            }),
        updateHackathonIdea: builder.mutation({
            query: (hackathonIdeaData) => ({
                url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                method: 'PUT',
                body: hackathonIdeaData.data,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
        deleteHackathonIdea: builder.mutation({
            query: ({hackathonId, ideaId}) => ({
                url: `hackathons/${hackathonId}/ideas/${ideaId}`,
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
    })
});

export const {
    useGetHackathonIdeasQuery,
    useCreateHackathonIdeaMutation,
    useDeleteHackathonIdeaMutation,
    useUpdateHackathonIdeaMutation
} = hackathonIdeaApi;
