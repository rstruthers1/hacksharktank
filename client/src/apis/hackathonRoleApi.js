import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonRoleApi = createApi({
  reducerPath: 'hackathonRoleApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
      getHackathonRoles: builder.query({
        query: () => ({
          url: `hackathon-roles`,
          method: 'GET',
          // Add a JWT token to the request headers if the user is logged in
          headers: {'Authorization': `JWT ${localStorage.getItem('token')}`},
        }),
        providesTags: (result, error, arg) => [{type: 'HackathonRoles', id: arg}],
      }),
    })

});

export const {useGetHackathonRolesQuery} = hackathonRoleApi;