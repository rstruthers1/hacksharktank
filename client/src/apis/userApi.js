// userApi.js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const userApi = createApi({
    reducerPath: 'userApi',
    baseQuery: fetchBaseQuery(
        { baseUrl: `${baseURL}/` }), // Set your base URL here
    endpoints: (builder) => ({
        registerUser: builder.mutation({
            query: (userData) => ({
                url: 'users',
                method: 'POST',
                body: userData,
            }),
        }),
        loginUser: builder.mutation({
            query: (userData) => ({
                url: 'users/login',
                method: 'POST',
                body: userData,
            }),
        }),
        searchUsers: builder.query({
            query: (searchTerm) => ({
                url: `users/search?searchTerm=${searchTerm}`,
                method: 'GET',
            }),
        }),
        getUser: builder.query({
            query: (userId) => ({
                url: `users/${userId}`,
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'User', id: arg }],
        }),
        updateUser: builder.mutation({
            query: (userData) => ({
                url: `users/${userData.id}`,
                method: 'PUT',
                body: userData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'User', id: arg.id }],
        }),
        // You can add more endpoints here
    }),
});

// Export the auto-generated hook for the `registerUser` and `loginUser` mutation
export const {
    useRegisterUserMutation,
    useLoginUserMutation,
    useLazySearchUsersQuery,
    useGetUserQuery,
    useUpdateUserMutation
} = userApi;
