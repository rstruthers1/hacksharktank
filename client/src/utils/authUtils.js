// authUtils.js
export const isUserLoggedIn = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Check if expired, else return true
    try {
        const {exp} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }

};

export const getLoggedInUser = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return null;
    }

    // Check if expired, else return true
    try {

        const {id, username, email} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        console.log(`id: ${id}, username: ${username}, email: ${email}`)
        return {id, email};
    } catch {
        return null; // if there's an error decoding the token, consider it invalid
    }
}

export const isUserAdmin = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }
    // Check if expired
    let user = null;
    try {
        user = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(user.exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration && user.roles.includes('admin');
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
}

// Returns true if token exists and is expired, else false
export const isSessionExpired = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Check if expired, else return true
    try {
        const {exp} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now > expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
}

export const logoutUser = () => {
    localStorage.removeItem('token');
}
