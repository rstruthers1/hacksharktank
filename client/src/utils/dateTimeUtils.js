const MILLIS_IN_ONE_DAY = 86400000;

export const getTodayAtMidnight = () => {
    const today = new Date();
    today.setHours(0,0,0,0);
    return today;
}

export const getMinEndDate = (startDate) => {
    if (!startDate) {
        return getTodayAtMidnight().getTime() + MILLIS_IN_ONE_DAY;
    }
    try {
        const startDateAsDate = new Date(startDate);
        const dayAfter = new Date(startDateAsDate.getTime() + MILLIS_IN_ONE_DAY);
        return dayAfter.getTime();
    } catch (err) {
        console.error(err);
        return getTodayAtMidnight().getTime() + MILLIS_IN_ONE_DAY;
    }
}

export const formatDateTime = (dateTime) => {
    if (!dateTime) {
        return '';
    }
    try {
        const dateTimeAsDate = new Date(dateTime);
        return dateTimeAsDate.toLocaleString();
    } catch (err) {
        console.error(err);
        return '';
    }
}

export const formatDateTimeAsDate  = (date) => {
    if (!date) {
        return '';
    }
    try {
        const dateAsDate = new Date(date);
        return dateAsDate.toLocaleDateString();
    } catch (err) {
        console.error(err);
        return '';
    }
}