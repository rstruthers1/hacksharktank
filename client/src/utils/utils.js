// utils.js
export const isUserLoggedIn = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Optionally check for token expiry if your JWT contains an expiry claim
    try {
        const { exp } = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
};
