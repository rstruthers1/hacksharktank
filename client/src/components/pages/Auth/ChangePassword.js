import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { Container, Form, Button } from 'react-bootstrap';
import { useChangePasswordMutation } from '../../../apis/userApi';
import {getLoggedInUser} from "../../../utils/authUtils";
import {toast} from "react-toastify";

// Yup schema for validation
const schema = yup.object({
  oldPassword: yup.string().required('Old password is required'),
  newPassword: yup.string().required('New password is required').min(8, 'Password must be at least 8 characters long')
        .matches(/[a-z]/, 'Password must contain at least one lowercase letter')
        .matches(/[A-Z]/, 'Password must contain at least one uppercase letter')
        .matches(/[0-9]/, 'Password must contain at least one number'),
  confirmNewPassword: yup.string()
      .oneOf([yup.ref('newPassword'), null], 'Passwords must match')
      .required('Confirm new password is required'),
}).required();

const ChangePassword = () => {
  const [changePassword, { isLoading, isSuccess, isError, data, error }] = useChangePasswordMutation();

  const { control, handleSubmit, formState: { errors } } = useForm({
    resolver: yupResolver(schema),
  });

  const onSubmit = async formData => {
    const userId = getLoggedInUser()?.id;
    const userData = {
      oldPassword: formData.oldPassword,
      newPassword: formData.newPassword,
      id: userId
    };
    await changePassword(userData).unwrap()
        .then(() => {
          toast.success("Password changed successfully");
        } )
        .catch((err) => {
          toast.error(`Error changing password: ${err.message}`);
        });
  };

  return (
      <Container>
        <h2>Change Password</h2>
        <Form onSubmit={handleSubmit(onSubmit)}>
          <Form.Group controlId="oldPassword">
            <Form.Label>Old Password</Form.Label>
            <Controller
                name="oldPassword"
                control={control}
                render={({ field }) => <Form.Control type="password" isInvalid={!!errors.oldPassword} {...field} />}
            />
            <Form.Control.Feedback type="invalid">
              {errors.oldPassword?.message}
            </Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="newPassword">
            <Form.Label>New Password</Form.Label>
            <Controller
                name="newPassword"
                control={control}
                render={({ field }) => <Form.Control type="password" isInvalid={!!errors.newPassword} {...field} />}
            />
            <Form.Control.Feedback type="invalid">
              {errors.newPassword?.message}
            </Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="confirmNewPassword">
            <Form.Label>Confirm New Password</Form.Label>
            <Controller
                name="confirmNewPassword"
                control={control}
                render={({ field }) => <Form.Control type="password" isInvalid={!!errors.confirmNewPassword} {...field} />}
            />
            <Form.Control.Feedback type="invalid">
              {errors.confirmNewPassword?.message}
            </Form.Control.Feedback>
          </Form.Group>

          <Button variant="primary" type="submit">
            Change Password
          </Button>
        </Form>
      </Container>
  );
};

export default ChangePassword;
