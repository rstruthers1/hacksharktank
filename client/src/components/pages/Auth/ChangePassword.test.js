import React from 'react';
import {render, screen, waitFor, fireEvent, getByLabelText} from '@testing-library/react';
import '@testing-library/jest-dom'
import ChangePassword from './ChangePassword'; // Adjust the import path as needed
import { useChangePasswordMutation } from '../../../apis/userApi';
import { getLoggedInUser } from "../../../utils/authUtils";

// Mocking dependencies
jest.mock('../../../apis/userApi', () => ({
    useChangePasswordMutation: jest.fn(),
}));

jest.mock('../../../utils/authUtils', () => ({
    getLoggedInUser: jest.fn(),
}));

jest.mock("react-toastify", () => ({
    toast: {
        success: jest.fn(),
        error: jest.fn(),
    },
}));

describe('<ChangePassword />', () => {

    beforeEach(() => {
        useChangePasswordMutation.mockReturnValue([jest.fn(), {}]);
        getLoggedInUser.mockReturnValue({ id: '123' });
    });

    it('renders the change password form', () => {
        render(<ChangePassword />);
        expect(screen.getByLabelText(/Old Password/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/^New Password/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/Confirm New Password/i)).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /Change Password/i })).toBeInTheDocument();
    });

    it('displays an error message when old password is not provided', async () => {
        render(<ChangePassword />);
        fireEvent.click(screen.getByRole('button', { name: /Change Password/i }));
        await waitFor(() => {
            expect(screen.getByText(/Old password is required/i)).toBeInTheDocument();
        });

    });

    // Add more test cases as needed, such as for validation errors, submission errors, etc.
});
