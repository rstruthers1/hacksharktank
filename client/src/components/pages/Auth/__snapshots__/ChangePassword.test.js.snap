// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePassword /> should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h2>
      Change Password
    </h2>
    <form
      class=""
    >
      <div>
        <label
          class="form-label"
          for="oldPassword"
        >
          Old Password
        </label>
        <input
          class="form-control"
          id="oldPassword"
          name="oldPassword"
          type="password"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div>
        <label
          class="form-label"
          for="newPassword"
        >
          New Password
        </label>
        <input
          class="form-control"
          id="newPassword"
          name="newPassword"
          type="password"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div>
        <label
          class="form-label"
          for="confirmNewPassword"
        >
          Confirm New Password
        </label>
        <input
          class="form-control"
          id="confirmNewPassword"
          name="confirmNewPassword"
          type="password"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </form>
  </div>
</div>
`;
